src/Hooks/useAuthActions.js

```diff
@@ -0,0 +1,149 @@
+import { auth, db } from "../Components/Firebase";
+import {
+  EmailAuthProvider,
+  GoogleAuthProvider,
+  TwitterAuthProvider,
+  createUserWithEmailAndPassword,
+  linkWithCredential,
+  linkWithPopup,
+  signInAnonymously,
+  signInWithEmailAndPassword,
+  signInWithPopup,
+} from "firebase/auth";
+import { useContext } from "react";
+import { LocalUserContext } from "../Components/LocalUserContext";
+import {
+  PopulateFromFirestore,
+  SaveToFirestore,
+} from "../Components/Firestore";
+import { doc, setDoc } from "firebase/firestore";
+
+export default function useAuthActions() {
+  const [localUser, setLocalUser] = useContext(LocalUserContext);
+
+  // A general function handling a successful authentication.
+  async function onSuccessImpl(
+    userCredential,
+    authProvider,
+    displayNameOverride,
+    isRegistration
+  ) {
+    const user = userCredential.user;
+
+    // Always write auth info to /users.
+    await setDoc(
+      doc(db, "users", user.uid),
+      {
+        uid: user.uid,
+        authProvider: authProvider,
+        email: user.email,
+      },
+      { merge: true }
+    );
+
+    // If this is registration, write onboarding LocalUser to /users.
+    if (isRegistration) {
+      // Set initial display name first.
+      const newLocalUser = { ...localUser };
+      newLocalUser.name = displayNameOverride ?? user.displayName;
+      setLocalUser(newLocalUser);
+      await SaveToFirestore(user, newLocalUser);
+    }
+
```

Comment:
The above SaveToFirestore will wipe an existing google/twitter account.  However, I see it's also the function that prevents the name stomping issue.

Perhaps, we add a second conditional statement, that only is truthy if the google/twitter user credentials are new to us.  I would like to query Firebase Authentication to confirm this....however, I can't figure it out rn. So my workaround would be querying Firestore instead.  (Note, while we can search for AuthProvider and Email for Google accounts, we would need to add in the TwitterHandle when registering Twitter accounts for this approach to work.

Also, I like the idea of throwing a snackbar to alert users if the account already exists, so they know their onboarding selections got scrapped.

Reply:
Hmm, very valid concerns.  I'm trying to make sure I understand the scenario where existing accounts could be wiped.  Are you considering a case where I am logged in using email/password, for example, then I go to register page and click 'log in with google' or something?  Because, in this case, it would seem like the LocalUser from the previous account would overwrite the new one, and that does seem bad.  Good catch.

I just added a commit that will silently log the user in to the existing account, in this case.  I check for existence by checking for the /users document, which seems like a pretty safe way to check this.  Take a look and let me know what you think. An alternative would be to fail a registration in this case -- I could do that easily as well, and show the user a message like "This identity already has an account.  Try logging in instead."  Not sure which is better.

Reply:
Actually, as I think about it, in the case that the user chooses an already existing account, it could go two ways:

1) The user was not already authenticated -- here, my new code would take effect and basically log the user in to that account, throwing away their onboarding data.

2) The user was already authenticated -- here, they will attempt to link the existing account with this google/twitter account.  This may not succeed if the existing account is not anonymous.  If it is anonymous, but they are trying to link with an identity that already has a different uid, Idk exactly what happens.  I *think* it should fail.  We should verify this, though...

Reply:
Haven't dug into the changes yet, but here's how to recreate the account wiping I noticed.  
1. Make an account using Google/Twitter and populate it. 
2. Sign out (I refresh too)
3. Go through onboarding and register an account using Google/Twitter.
4. All info in account from Step 1 will be wiped.